test/Remittance.js

```diff
@@ -0,0 +1,240 @@
+            assert.strictEqual(txObj.logs[0].args[2].toString(), toWei("0.1", "ether"));
+    
+            txObj = await instance.newTransaction(exchange, password, {from: owner, value: toWei("0.1", "ether")});
+            assert.strictEqual(txObj.logs.length, 1);
+            assert.strictEqual(txObj.logs[0].event, "LogNewTransaction");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1], exchange);
+            assert.strictEqual(fromWei(txObj.logs[0].args[2], "ether"), "0.1");        
+        });
+    
+        it('Should not be able to cancel a transaction if the sender is not the tx emitter', async () => {
+            await increaseTime(5 * 24 * 60 * 60);
+    
+            await truffleAssert.reverts(
+                instance.cancelTransaction(exchange, password, {from: stranger})
+            ); 
+        });
+    
+        it('Should not be able to cancel a transaction with invalid address & password', async () => {
+            await increaseTime(5 * 24 * 60 * 60);
+    
+            await truffleAssert.reverts(
+                instance.cancelTransaction(exchange, web3.utils.utf8ToHex(""), {from: stranger})
+            );
+            await truffleAssert.reverts(
+                instance.cancelTransaction(nil, password, {from: owner})
+            ); 
+        });
+    
+        it('Should not be able to cancel a transaction before the deadline', async () => {
+            await truffleAssert.reverts(
+                instance.cancelTransaction(exchange, password, {from: owner})
+            );
+        });
+    });
+
+    describe('======= withdraw unit testing =======', () => {
+        beforeEach('Creating new transaction', async () => {
+            const txObj = await instance.newTransaction(exchange, password, {from: owner, value: toWei("0.1", "ether")});
+
+            assert.strictEqual(txObj.logs.length, 1);
+            assert.strictEqual(txObj.logs[0].event, "LogNewTransaction");
+            assert.strictEqual(txObj.logs[0].args[0], owner);
+            assert.strictEqual(txObj.logs[0].args[1], exchange);
+            assert.strictEqual(fromWei(txObj.logs[0].args[2], "ether"), "0.1");
+        });
+        
+        it('Should fail if the provided password is invalid', async () => {
+            await truffleAssert.reverts(
+                instance.withdraw(web3.utils.utf8ToHex("amIAwes0m3?"), {from: exchange})
+            );
+        });
+
+        it('Should fail if the sender is not the stated exchange', async () => {
+            await truffleAssert.reverts(
+                instance.withdraw(password, {from: stranger})
+            );
+        });
+
+        it('Should withdraw the exchange balance', async () => {
+            let initialBalance = new BN(await web3.eth.getBalance(exchange));
+
+            const txObj = await instance.withdraw(password, {
+                from: exchange,
+                gasPrice: 50
```

Comment:
Have we seen how to do it without imposing the gas price?

Reply:
I must admit that I don't remember this subtility

Reply:
You let the system choose the gas price for you, then, at calculation time, you look at what was picked for you.